File: ui/apps/web/src/lib/api/position.ts
/** Where a download was left in the player, kept by the API so it resumes on any device (#96). */

import { putApi } from "./client";
import { normalizePosition, type PositionView } from "./task";

/** How often the player saves while it plays, as well as on pause and on close. */
export const POSITION_SAVE_INTERVAL_MS = 10_000;

export async function savePosition(
    taskId: string,
    fileIndex: number | null,
    positionSeconds: number,
    durationSeconds: number,
): Promise<PositionView> {
    return normalizePosition(
        await putApi<any>(`/download/${taskId}/position`, {
            ...(fileIndex === null ? {} : { file_index: fileIndex }),
            position_seconds: positionSeconds,
            duration_seconds: durationSeconds,
        }),
    );
}

File: api/src/data/db/model/download/position.py
from __future__ import annotations

from typing import ClassVar

from tortoise import fields

from src.core.base import LinkBase


class PlaybackPosition(LinkBase):
    """Where a download was left in the player, so it resumes on any device (#96).

    One row per task and file: ``file_index`` is a torrent's file, and 0 for a
    download's one file. Never NULL, because Postgres counts NULLs as distinct
    and the pair would stop being unique.

    ``watched`` outlives the position: stopping near the end clears where to
    resume, and the card still counts the file as seen.
    """

    # Not "positions": ``TaskSchema`` has a field of that name, and a relation
    # of the same name shadows it with Tortoise's manager (see ``File``).
    task = fields.ForeignKeyField(
        "model.Task", related_name="playback_positions", on_delete=fields.CASCADE
    )
    file_index: int = fields.IntField(default=0)
    position_seconds: float = fields.FloatField(default=0.0)
    duration_seconds: float = fields.FloatField(default=0.0)
    watched: bool = fields.BooleanField(default=False)

    class Meta:
        table: ClassVar[str] = "playback_position"
        table_description: ClassVar[str] = "PlaybackPosition"
        ordering: ClassVar[list[str]] = ["file_index"]
        unique_together: ClassVar[tuple[tuple[str, ...], ...]] = (("task", "file_index"),)
